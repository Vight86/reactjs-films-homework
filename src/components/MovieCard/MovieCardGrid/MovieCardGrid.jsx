import React from 'react';
import propTypes from 'prop-types';
import { Link, useRouteMatch } from 'react-router-dom';
import Button from '../../Button/index';
import Title from '../../Title/index';
import GenreList from '../../GenreList/index';
import Rating from '../../Rating/index';
import Cover from '../../Cover/index';
import MovieSynopsis from '../../MovieSynopsis/index';
import styles from './MovieCardGrid.scss';

const MovieCardGrid = ({ data }) => {
  const {
    movie,
    genres,
    isInfoShown,
    handleShowInfo,
    isModalOpened,
    handleModalToggle,
  } = data;

  const {
    id, title, voteAverage, posterPath, backdropPath, genreIds, overview,
  } = movie;

  const genreList = genreIds.map((genreId) => genres.find((genre) => genre.id === genreId)?.name);

  const url = useRouteMatch()?.url;

  return (
    <article className={styles.movieCard}>

      <section className={styles.poster}>
        <Cover className="inCard">
          {posterPath && <img src={`https://image.tmdb.org/t/p/w780${posterPath}`} alt={title} />}
        </Cover>
        <div className={styles.viewInfo}>
          <Button className="playButton" onClick={() => handleModalToggle(id, isModalOpened)} />
          <span className={styles.watchNowText}>Watch Now</span>
          <Button className="primary" onClick={handleShowInfo}>View Info</Button>
        </div>
      </section>

      <section className={styles.generalInfo}>
        <Title className="secondary">
          <Link to={`${url}/movie/${id}`}>
            {title}
          </Link>
        </Title>
        <GenreList className="secondary" order={3}>{genreList}</GenreList>
        <Rating className="secondary" withStars={false}>{voteAverage}</Rating>
      </section>

      <section
        className={`${styles.infoBlock} ${isInfoShown === true ? styles.infoBlock_active : ''}`.trim()}
      >
        <Cover className="backdropCover">
          {backdropPath && <img src={`https://image.tmdb.org/t/p/w780${posterPath}`} alt={title} />}
        </Cover>
        <Button className="closeButton_inMovieCard" onClick={handleShowInfo}>&times;</Button>

        <section className={styles.generalInfo}>
          <Title className="tertiary">
            <Link to={`${url}/movie/${id}`}>
              {title}
            </Link>
          </Title>
          <GenreList className="secondary" order={3}>{genreList}</GenreList>
          <Rating className="secondary" withStars={false}>{voteAverage}</Rating>
        </section>

        <MovieSynopsis className="secondary_grid" isShown={false}>{overview}</MovieSynopsis>
        <Button className="secondary" onClick={() => handleModalToggle(id, isModalOpened)}>Watch Now</Button>
      </section>

    </article>
  );
};

MovieCardGrid.propTypes = {
  data: propTypes.shape({
    movie: propTypes.shape({
      adult: propTypes.bool,
      backdropPath: propTypes.string,
      genreIds: propTypes.arrayOf(propTypes.number),
      id: propTypes.number,
      originalLanguage: propTypes.string,
      originalTitle: propTypes.string,
      overview: propTypes.string,
      popularity: propTypes.number,
      posterPath: propTypes.string,
      releaseDate: propTypes.string,
      title: propTypes.string,
      video: propTypes.bool,
      voteAverage: propTypes.number,
      voteCount: propTypes.number,
    }),
    genres: propTypes.arrayOf(propTypes.object),
    isInfoShown: propTypes.bool,
    handleShowInfo: propTypes.func,
    isModalOpened: propTypes.bool,
    handleModalToggle: propTypes.func,
  }).isRequired,
};

export default MovieCardGrid;
