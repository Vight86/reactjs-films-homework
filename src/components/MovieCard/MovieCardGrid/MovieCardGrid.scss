@import '../../../utils/variables';

.viewInfo {
  position: absolute;
  top: 0%;
  left: 0;
  width: 100%;
  height: 100%;
  padding-top: 21%;
  background-color: transparentize($color: $color-dark, $amount: 0.16);
  opacity: 0;
  text-align: center;
  transform: scale(1.25);
  transition: all 0.3s;
}

.movieCard {
  position: relative;
  overflow: hidden;
  padding-bottom: 3.6em;
  background-color: $color-light;
  border-radius: 0.6em;
  box-shadow: 0 10px 40px 0 rgba(0, 0, 0, 0.04);
  font-size: 1rem;

  &:hover {
    .viewInfo {
      opacity: 1;
      transform: scale(1);
    }
  }

  @media screen and (max-width: $bp-medium) {
    font-size: 0.8rem;
  }

  @media screen and (max-width: $bp-xsmall) {
    font-size: 1.2rem;
  }

  @media screen and (max-width: $bp-xxsmall) {
    font-size: 0.8rem;
  }

  @media screen and (max-width: $bp-xxxsmall) {
    font-size: 1.2rem;
  }

  @media screen and (max-width: $bp-xxxxsmall) {
    font-size: 0.8rem;
  }
}

.poster {
  position: relative;
  overflow: hidden;
  width: 100%;
  margin: 0 0 5.5em 0;
}

.watchNowText {
  display: block;
  margin-bottom: 11%;
  color: $bg-color;
  font-size: 2.8em;
  letter-spacing: 0.04em;
}

.generalInfo {
  position: relative;
  z-index: 15;
  display: grid;
  padding: 0 8% 0 8%;
  grid-template-columns: 1fr 20%;
  grid-template-rows: 1fr 1fr;
}

.infoBlock {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  padding-top: 24.3%;
  background-color: $color-tertiary;
  background-position: center;
  background-repeat: no-repeat;
  background-size: auto, cover;
  opacity: 0;
  transition: all 0.5s;
  visibility: hidden;

  &::after {
    position: absolute;
    z-index: 7;
    top: 0;
    left: 0;
    display: block;
    width: 100%;
    height: 100%;
    background-color: transparentize($color: $color-dark, $amount: 0.36);
    content: '';
  }

  &_active {
    z-index: 20;
    opacity: 1;
    visibility: visible;
  }
}
