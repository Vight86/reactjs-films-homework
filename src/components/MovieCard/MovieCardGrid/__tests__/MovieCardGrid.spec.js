import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import MovieCardGrid from '../index';

let mockUseRouteMatchObject;
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useRouteMatch: jest.fn(() => mockUseRouteMatchObject),
  Link: jest.fn(),
}));

let mockHandleModalToggle;
let mockData;
let renderer;
let result;

beforeEach(() => {
  mockUseRouteMatchObject = { url: '/tested' };
  mockHandleModalToggle = jest.fn(() => 'tested');

  mockData = {
    movie: {
      adult: false,
      backdropPath: '/test.jpg',
      genreIds: [1, 2],
      id: 1,
      originalLanguage: 'en',
      originalTitle: 'test',
      overview: 'test',
      popularity: 1,
      posterPath: '/test.jpg',
      releaseDate: '1',
      title: 'test',
      video: false,
      voteAverage: 1,
      voteCount: 1,
    },
    genres: [{
      id: 1,
      name: 'Action',
    },
    {
      id: 2,
      name: 'SciFi',
    }],
    movieTrailerKey: 'test',
    isInfoShown: false,
    handleShowInfo() { },
    isModalOpened: false,
    handleModalToggle() {
      mockHandleModalToggle();
    },
  };

  renderer = new ShallowRenderer();
  renderer.render(<MovieCardGrid data={mockData} />);
  result = renderer.getRenderOutput();
});

afterEach(() => {
  mockHandleModalToggle.mockClear();
  mockData = null;
  renderer = null;
  result = null;
});

describe('render MovieCardGrid component', () => {
  it('render correctly', () => {
    expect(result).toMatchSnapshot();
  });

  it('render playButton click', () => {
    const playButton = result.props.children[0].props.children[1].props.children[0];
    playButton.props.onClick();
    result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('render watchNow button click', () => {
    const watchNowButton = result.props.children[2].props.children[4];
    watchNowButton.props.onClick();
    result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('render viewInfoButton button click', () => {
    let infoBlock = result.props.children[2].props.className;
    expect(infoBlock).toMatch(/infoBlock/);
    mockData.isInfoShown = true;
    renderer.render(<MovieCardGrid data={mockData} />);
    result = renderer.getRenderOutput();
    infoBlock = result.props.children[2].props.className;
    expect(infoBlock).toMatch(/infoBlock_active/);
  });
});
