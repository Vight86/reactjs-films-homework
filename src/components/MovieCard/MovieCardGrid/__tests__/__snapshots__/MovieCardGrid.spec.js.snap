// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render MovieCardGrid component render correctly 1`] = `
<article
  className="movieCard"
>
  <section
    className="poster"
  >
    <Cover
      className="inCard"
    >
      <img
        alt="test"
        src="https://image.tmdb.org/t/p/w780/test.jpg"
      />
    </Cover>
    <div
      className="viewInfo"
    >
      <Button
        className="playButton"
        onClick={[Function]}
      >
        
      </Button>
      <span
        className="watchNowText"
      >
        Watch Now
      </span>
      <Button
        className="primary"
        onClick={[Function]}
      >
        View Info
      </Button>
    </div>
  </section>
  <section
    className="generalInfo"
  >
    <Title
      className="secondary"
    >
      <mockConstructor
        to="/tested/movie/1"
      >
        test
      </mockConstructor>
    </Title>
    <GenreList
      className="secondary"
      order={3}
    >
      Action
      SciFi
    </GenreList>
    <Rating
      className="secondary"
      withStars={false}
    >
      1
    </Rating>
  </section>
  <section
    className="infoBlock"
  >
    <Cover
      className="backdropCover"
    >
      <img
        alt="test"
        src="https://image.tmdb.org/t/p/w780/test.jpg"
      />
    </Cover>
    <Button
      className="closeButton_inMovieCard"
      onClick={[Function]}
    >
      ×
    </Button>
    <section
      className="generalInfo"
    >
      <Title
        className="tertiary"
      >
        <mockConstructor
          to="/tested/movie/1"
        >
          test
        </mockConstructor>
      </Title>
      <GenreList
        className="secondary"
        order={3}
      >
        Action
        SciFi
      </GenreList>
      <Rating
        className="secondary"
        withStars={false}
      >
        1
      </Rating>
    </section>
    <MovieSynopsis
      className="secondary_grid"
      isShown={false}
    >
      test
    </MovieSynopsis>
    <Button
      className="secondary"
      onClick={[Function]}
    >
      Watch Now
    </Button>
  </section>
</article>
`;

exports[`render MovieCardGrid component render playButton click 1`] = `
<article
  className="movieCard"
>
  <section
    className="poster"
  >
    <Cover
      className="inCard"
    >
      <img
        alt="test"
        src="https://image.tmdb.org/t/p/w780/test.jpg"
      />
    </Cover>
    <div
      className="viewInfo"
    >
      <Button
        className="playButton"
        onClick={[Function]}
      >
        
      </Button>
      <span
        className="watchNowText"
      >
        Watch Now
      </span>
      <Button
        className="primary"
        onClick={[Function]}
      >
        View Info
      </Button>
    </div>
  </section>
  <section
    className="generalInfo"
  >
    <Title
      className="secondary"
    >
      <mockConstructor
        to="/tested/movie/1"
      >
        test
      </mockConstructor>
    </Title>
    <GenreList
      className="secondary"
      order={3}
    >
      Action
      SciFi
    </GenreList>
    <Rating
      className="secondary"
      withStars={false}
    >
      1
    </Rating>
  </section>
  <section
    className="infoBlock"
  >
    <Cover
      className="backdropCover"
    >
      <img
        alt="test"
        src="https://image.tmdb.org/t/p/w780/test.jpg"
      />
    </Cover>
    <Button
      className="closeButton_inMovieCard"
      onClick={[Function]}
    >
      ×
    </Button>
    <section
      className="generalInfo"
    >
      <Title
        className="tertiary"
      >
        <mockConstructor
          to="/tested/movie/1"
        >
          test
        </mockConstructor>
      </Title>
      <GenreList
        className="secondary"
        order={3}
      >
        Action
        SciFi
      </GenreList>
      <Rating
        className="secondary"
        withStars={false}
      >
        1
      </Rating>
    </section>
    <MovieSynopsis
      className="secondary_grid"
      isShown={false}
    >
      test
    </MovieSynopsis>
    <Button
      className="secondary"
      onClick={[Function]}
    >
      Watch Now
    </Button>
  </section>
</article>
`;

exports[`render MovieCardGrid component render watchNow button click 1`] = `
<article
  className="movieCard"
>
  <section
    className="poster"
  >
    <Cover
      className="inCard"
    >
      <img
        alt="test"
        src="https://image.tmdb.org/t/p/w780/test.jpg"
      />
    </Cover>
    <div
      className="viewInfo"
    >
      <Button
        className="playButton"
        onClick={[Function]}
      >
        
      </Button>
      <span
        className="watchNowText"
      >
        Watch Now
      </span>
      <Button
        className="primary"
        onClick={[Function]}
      >
        View Info
      </Button>
    </div>
  </section>
  <section
    className="generalInfo"
  >
    <Title
      className="secondary"
    >
      <mockConstructor
        to="/tested/movie/1"
      >
        test
      </mockConstructor>
    </Title>
    <GenreList
      className="secondary"
      order={3}
    >
      Action
      SciFi
    </GenreList>
    <Rating
      className="secondary"
      withStars={false}
    >
      1
    </Rating>
  </section>
  <section
    className="infoBlock"
  >
    <Cover
      className="backdropCover"
    >
      <img
        alt="test"
        src="https://image.tmdb.org/t/p/w780/test.jpg"
      />
    </Cover>
    <Button
      className="closeButton_inMovieCard"
      onClick={[Function]}
    >
      ×
    </Button>
    <section
      className="generalInfo"
    >
      <Title
        className="tertiary"
      >
        <mockConstructor
          to="/tested/movie/1"
        >
          test
        </mockConstructor>
      </Title>
      <GenreList
        className="secondary"
        order={3}
      >
        Action
        SciFi
      </GenreList>
      <Rating
        className="secondary"
        withStars={false}
      >
        1
      </Rating>
    </section>
    <MovieSynopsis
      className="secondary_grid"
      isShown={false}
    >
      test
    </MovieSynopsis>
    <Button
      className="secondary"
      onClick={[Function]}
    >
      Watch Now
    </Button>
  </section>
</article>
`;
